src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -311,12 +342,57 @@ public static RoutePattern Pattern(string rawText, params RoutePatternPathSegmen
                 }
             }
 
+            // Each Required Value either needs to either:
+            // 1. be null-ish
+            // 2. have a corresponding parameter
+            // 3. have a corrsponding default that matches both key and value
+            if (requiredValues != null)
+            {
+                foreach (var kvp in requiredValues)
+                {
+                    // 1.be null-ish
+                    var found = RouteValueEqualityComparer.Default.Equals(string.Empty, kvp.Value);
+
+                    // 2. have a corresponding parameter
+                    if (!found && parameters != null)
+                    {
+                        for (var i = 0; i < parameters.Count; i++)
+                        {
+                            if (string.Equals(kvp.Key, parameters[i].Name, StringComparison.OrdinalIgnoreCase))
+                            {
+                                found = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    // 3. have a corrsponding default that matches both key and value
+                    if (!found && 
+                        updatedDefaults != null && 
+                        updatedDefaults.TryGetValue(kvp.Key, out var defaultValue) &&
+                        RouteValueEqualityComparer.Default.Equals(kvp.Value, defaultValue))
+                    {
+                        found = true;
+                    }
+
+                    if (!found)
+                    {
+                        throw new InvalidOperationException(
+                            $"No corresponding parameter or default value could be found for the required value " +
+                            $"'{kvp.Key}={kvp.Value}'. A non-null required value must correspond to a route parameter or the " +
+                            $"route pattern must have a matching default value.");
+                    }
+                }
+            }
+
+
             return new RoutePattern(
                 rawText,
                 updatedDefaults ?? EmptyDefaultsDictionary,
                 updatedParameterPolicies != null
                     ? updatedParameterPolicies.ToDictionary(kvp => kvp.Key, kvp => (IReadOnlyList<RoutePatternParameterPolicyReference>)kvp.Value.ToArray())
                     : EmptyPoliciesDictionary,
+                requiredValues ?? EmptyDefaultsDictionary,
```

Comment:
Rename EmptyDefaultsDictionary to something more generic

Reply:
sure